src/rust/async-cache/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+        // queue and querying the plugin-registry for updates corresponding to
+        // each message.
+        tokio::task::spawn(async move {
+            updater_rx
+                .for_each_concurrent(updater_pool_size, move |key: K| {
+                    let cache = cache.clone();
+                    let mut updater = updater.clone();
+
+                    async move {
+                        let value = updater(key.clone()).await;
+                        cache.insert(key, value).await;
+                    }
+                })
+                .await;
+        });
+
+        Self {
+            cache: cache_retval,
+            updater_tx,
+        }
+    }
+
+    /// Retrieve the value corresponding to the given key.
+    ///
+    /// This method attempts to resolve the request against a local cache in
+    /// memory. Upon cache miss, we attempt to enqueue a cache update and return
+    /// immediately.
+    ///
+    /// # Arguments
+    ///
+    /// * key - the key to retrieve a value for
+    ///
+    /// # Returns
+    ///
+    /// * Ok(Some(value)) - A value if there is an entry present in the cache.
+    ///
+    /// * Ok(None) - No entry corresponding to the given key was found in the
+    ///   cache, but an update request was successfully enqueued.
+    ///
+    /// * Err(e) - No entry corresponding to the given key was found in the
+    ///   cache, and we failed to enqueue an update request.  Failures take two
+    ///   forms:
+    ///
+    ///   - AsyncCacheError::Retryable - e.g. we failed to enqueue an update
+    ///     request but it should work eventually if retried.
+    ///
+    ///   - AsyncCacheError::Fatal - something happened which has "poisoned" the
+    ///     AsyncCache instance such that all future calls to this method will
+    ///     fail.
+    //#[tracing::instrument(skip(self, key))]
+    pub async fn get(&mut self, key: K) -> Result<Option<V>, AsyncCacheError> {
+        if let Some(value) = self.cache.get(&key) {
+            Ok(Some(value)) // cache hit
+        } else if let Err(e) = self.updater_tx.try_send(key) {
+            Err(e.into()) // cache miss, failed to enqueue an update
+        } else {
+            Ok(None) // cache miss, successfully enqueued update
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use super::AsyncCache;
+
+    #[tokio::test]
+    async fn test_cache_hit() {
+        let mut async_cache =
+            AsyncCache::new(10, Duration::from_millis(100), 10, 10, |_| async { 42 }).await;
+
+        let foo = async_cache.get(3).await.expect("should be successful");
+        assert_eq!(foo, None); // empty but update has been enqueued
+
+        tokio::time::sleep(Duration::from_millis(1)).await; // "short delay"
+
+        let foo = async_cache.get(3).await.expect("should be successful");
+        assert_eq!(foo, Some(42)); // cache updated and value is present
```

Comment:
I know this is pre-existing code you just moved to its own lib - great, btw - but, is there any guarantee that the `updater_rx` will have received and followed through with updating for key 3 after 1ms, or is it a potentially-race-condition?

Reply:
It's subject to the async runtime, so there's no guarantee. In practice 1ms appears to be long enough.